Implement horizontal scrolling

Guldoman opened the pull request:
This allows horizontal scrolling via `shift+scroll`.
For now only `DocView` can scroll, but anything that extends `View` should be able to horizontally scroll by returning a second value from `get_scrollable_size`.

Most of the complexity of this PR is due to the need to keep track of the longest lines. This avoids re-scanning the entire document every time it changes. Instead only altered lines are checked.
The only occasion that a full re-scan is required, is when there is only one longest line and it got shortened.

Fixes #424.

takase1121 commented on data/core/view.lua:
```diff
@@ -121,8 +126,12 @@ end
-  local max = self:get_scrollable_size() - self.size.y
-  self.scroll.to.y = common.clamp(self.scroll.to.y, 0, max)
+  local scrollsize_y, scrollsize_x = self:get_scrollable_size()
+  scrollsize_x = scrollsize_x or 0 -- FIXME: not every subclass returns the second value
```
I think it is fine to leave it as `math.huge`. It'll at least allow users to try to scroll instead of locking them.

Guldoman replied:
I left it at `0` to avoid changing the behavior of other classes. Changing it to `math.huge` has amusing effects like scrolling the command palette.

takase1121 commented:
Sorry for this ignorant question, but why is the longest line stored in a table instead of just a number? I read through the code and really can't figure it out.

Guldoman commented:
I did it to avoid having to fully re-scan the document if one of the longest lines is shortened. With the table we know if there are other lines with the same longest length.

takase1121 commented:
If you don't mind, can you also add a horizontal scrollbar?

franko commented:
> If you don't mind, can you also add a horizontal scrollbar?

This was also my thought.

Lite XL aims to have its functional elements apparent and visible to the user so that people can figure out things without having to google or reading a manual page. This applies also to "states" in the sense that when the application is in a specific transient state it should be visible to the user.

Guldoman commented:
Yeah, sure. I'll try to see what I can do later.

Guldoman commented:
I had to add some code to work around compatibility problems with some plugins. For example the `minimap` plugin overrides some `View`/`DocView` functions.

When possible I would like to remove these workarounds and update those plugins.

franko commented:
I made a test locally and I like this very much, thank you @Guldoman.

I will just wait a little bit before merging to let other contributors have a look and also for myself to review the code in detail.

franko commented on data/core/doc/init.lua:
```diff
@@ -325,6 +339,20 @@ function Doc:raw_insert(line, col, text, undo_stack, time)
 
+  if #lines > 1 then
+    -- Need to shift all the subsequent long lines
+    local line_numbers = {}
```
I think we may avoid recreating a new table here for line_numbers.

Guldoman replied:
How should we do that? We need to modify entries in the table that follow the current one.

franko commented on data/core/doc/init.lua:
```diff
@@ -499,6 +544,34 @@ function Doc:indent_text(unindent, line1, col1, line2, col2)
+  local max_length = self.long_lines.length
+  end_line = math.min(end_line, #self.lines)
+
+  for line=start_line,end_line do
```
We may use the same approach I suggested above to add items in the line and remove the spurious entries at the end. We have the problem that the value for each key entry is "true" and not equal to its len. Actually we may leave the values in the table with the length of the line instead of using "true"s values.

franko commented on data/core/doc/init.lua:
```diff
@@ -499,6 +544,34 @@ function Doc:indent_text(unindent, line1, col1, line2, col2)
+    end
+  end
+
+  if common.get_table_size(line_numbers) == 0 then
```
The function `common.get_table_size` should be removed. To check if a table has no elements the condition `next(t) == nil` can be used.

Guldoman replied:
Well, that's handy. I didn't know about it, I still need to research lua more.

franko commented on data/core/doc/init.lua:
```diff
@@ -499,6 +544,34 @@ function Doc:indent_text(unindent, line1, col1, line2, col2)
+  if common.get_table_size(line_numbers) == 0 then
+    -- Recalc needed
+    self.long_lines.length = 0
+    self.long_lines.line_numbers = { [1] = true }
```
In `line_numbers` was empty why creating a new table object ?

franko commented on data/core/docview.lua:
```diff
@@ -97,7 +97,16 @@ end
+  end
+  local size_v = self:get_line_height() * (#self.doc.lines - 1) + self.size.y
+  local size_h = self:get_col_x_offset(long_line, self.doc.long_lines.length)
+                 + self.size.x - (self.size.x - self:get_gutter_width()) + xmargin
```
+ self.size.x - self.size.x ?

Guldoman replied:
It was to keep a similar style of `+ self.size.y` that then gets subtracted in `clamp_scroll_position`. But yeah, it looks redundant, I'll remove it.

franko commented on data/core/view.lua:
```diff
@@ -46,51 +47,77 @@ function View:get_name()
 
-
+-- Returns `y, x` for compatibility
 function View:get_scrollable_size()
```
I think we are abusing this function here and we may break compatibility. Shouldn't we use another function for horizontal scrolling like get_h_scrollable_size ?

franko commented on data/core/view.lua:
```diff
@@ -46,51 +47,77 @@ function View:get_name()
 function View:scrollbar_overlaps_point(x, y)
-  local sx, sy, sw, sh = self:get_scrollbar_rect()
-  return x >= sx - sw * 3 and x < sx + sw and y >= sy and y < sy + sh
+  local v, h = self:get_scrollbars_rect()
```
Shouldn't we have a separate function for the horizontal axis ?

franko commented on data/core/view.lua:
```diff
@@ -46,51 +47,77 @@ function View:get_name()
-    return true
-  end
+  self.dragging_v_scrollbar, self.dragging_h_scrollbar = self:scrollbar_overlaps_point(x, y)
+  self.dragging_scrollbar = self.dragging_v_scrollbar -- TODO: make plugins use `self.dragging_v_scrollbar`
```
I advice to not touch the plugins and keep `self.dragging_scrollbar` just for the vertical scrolling. The rule will be that scrolling without "h" in their name will be about vertical scrolling.

I suggest that in the code we drop `self.dragging_v_scrollbar` in favor of `self.dragging_scrollbar`. I recognize we violate the symmetry between h and v but v came first, historically.

franko reviewed (commented):
I made a lot of comments, please have a look.

Broadly they are of two kinds:

- let's avoid creating always new objects for line_numbers if possible
- let's not modify existing functions and create separate one for h scrolling

For the plugins, they should not require any modification. When they call function about scrolling it will be implicitly vertical scrolling and any function about h scrolling should be explicit.

Guldoman commented on data/core/doc/init.lua:
```diff
@@ -60,19 +61,32 @@ end
+    local line_len = string.len(line) + 1 -- account for newline
+    if line_len > max_length then
+      max_length = line_len
+      line_numbers = { [i] = true } -- new longest line
```
This is very nice, I like this.

Guldoman commented:
I split everything about scrollbars in horizontal and vertical, and modified the code to be more in line of how it was before. Let me know if I missed anything.

franko commented on data/core/view.lua:
```diff
@@ -66,31 +72,59 @@ function View:get_scrollbar_rect()
 function View:scrollbar_overlaps_point(x, y)
   local sx, sy, sw, sh = self:get_scrollbar_rect()
-  return x >= sx - sw * 3 and x < sx + sw and y >= sy and y < sy + sh
+  return x >= sx - sw * 3 and x < sx + sw and y >= sy and y <= sy + sh
```
Here you modified the expression. It looks to me like an accidental modification.

Guldoman replied:
It it done on purpose; there was a pixel at the end of the scrollbar that wasn't being accounted for.

franko replied:
Fine for me.

franko commented on data/core/view.lua:
```diff
@@ -66,31 +72,59 @@ function View:get_scrollbar_rect()
+
+function View:h_scrollbar_overlaps_point(x, y)
+  local sx, sy, sw, sh = self:get_h_scrollbar_rect()
+  return x >= sx and x <= sx + sw and y > sy - sh * 3 and y <= sy + sh
```
Like comment above: there is a reason why you are not using y < sy + sh ? I don't mean it is wrong, just asking.

Guldoman replied:
Same reason of the vertical one. A pixel at the bottom of the scrollbar not accounted for.

franko replied:
Fine.

franko commented on data/core/doc/init.lua:
```diff
@@ -499,6 +545,35 @@ function Doc:indent_text(unindent, line1, col1, line2, col2)
+  if not next(line_numbers) then
+    -- Recalc needed
+    self.long_lines.length = 0
+    self.long_lines.line_numbers = { }
```
Why creating a new `line_numbers` if the previous one is already empty ?

Guldoman replied:
I'll fix this in the next commit.

franko replied:
Ok, I will merge as soon as this is done even if this isn't terribly important. Thank you for your patience.

franko reviewed (commented):
Thank you for integrating my previous comments. Now it looks fine, I just made a small comments you may check but otherwise I am okay to merge. I made also some tests and everything works beautifully.

franko commented:
Just a note about an idea I wanted to share with you. I don't mean it should be implemented in this PR but we may implement in a further modification if it is sound and we want to proceed.

You are saving the `long_lines.line_numbers` as a table whose keys are the line numbers. Instead I thought we could save it as a list of line numbers, so, for example {13, 87, 101} instead of {{13] = true, [87] = true, [101] = true}.

Now, we would gain very little or nothing with this change but we may now proceed in this way: instead of using the line numbers directly we could use just the difference in the line number from the previous entry. So instead of {13, 87 - 13, 101 - 87} or explicitly {13, 74, 14}. The advantage should be that on raw_insert/remove it should be easier to modify with less operations and no need of creating a new table. On the other side, when iterating to have the line numbers it would be easy to determine the line number just by keeping a variable pointing to the line number of the previous entry.

For the initial loop on document open the loop could be:

```lua
  local ill = 0
  for line in fp:lines() do
    ...
    local line_len = #line + 1 -- account for newline
    if line_len > max_length then
      ill = 1
      line_numbers[ill] = line_len
    elseif line_len == max_length then
      ill = ill + 1
      line_numbers[ill] = line_len
    end
    i = i + 1
  end
  -- cleanup spurious entries in line_numbers
  while ill < #line_numbers do table.remove(line_numbers) end
```

franko commented on data/core/doc/init.lua:
```diff
@@ -499,6 +545,36 @@ function Doc:indent_text(unindent, line1, col1, line2, col2)
+  if not next(line_numbers) then
+    -- Recalc needed
+    self.long_lines.length = 0
+    self.long_lines.line_numbers = line_numbers
```
This assignment does nothing I think because the lh side was already equal to the rh side but that's ok.

franko commented:
Thank you again for this work, well done!

Guldoman commented:
> Just a note about an idea I wanted to share with you. I don't mean it should be implemented in this PR but we may implement in a further modification if it is sound and we want to proceed.

Yeah, in the next days I'll think more about how to better optimize this code. The main bottleneck is having to sometimes recalculate the longest lines, so a way to avoid that would be optimal.
